### PR TITLE
Upgrade text-numeric inferring

### LINKED ISSUE #1107: Handle more human readable number imports
## Problem

Currently, we only detect a column as a number type if the entries can be cast to a `NUMERIC` using the default PostgreSQL casting functionality.  This is a problem in certain locales, and also can't handle numbers with grouping separators (e.g., the `,` in `1,000.00`).

## Proposed solution

We should write custom logic to allow handling more strings when casting to number types.

## Additional context

- This is inspired by and related to the custom casting work we're doing for money types. Please see PR #1137 for related code.

### PR DESCRIPTION
**Pull request by Jyuart**
Fixes #1107

It's an upgrade of the text (`TEXT`, `CHAR`, `VARCHAR`) --> numeric (`NUMERIC`) types inferring (and more comprehensive casting to numeric type accordingly since inferring is based on casting capabilities). Values like:
- 3.14
- 123,456.7
- 123.456,7
- 123 456,7
- 1,23,456.7
- 123'456.7
are now successfully recognized as of `NUMERIC` type

**Technical details**

It was inspired by this PR https://github.com/centerofci/mathesar/pull/1137 and is implemented in the same manner:
1. A separate function for getting a numeric array in the format: `['number', 'separator', 'floating-point']`. For example: `['331,209.05', ',', '.']`
2. Another function that removes locale-specific parts and extracts the actual number in the format with `.` as a floating-point and with no separators: `331209.05`
3. Supported formats are based on this comment: https://github.com/centerofci/mathesar/issues/1107#issuecomment-1123949497, but can be easily upgraded if necessary. Excerpt:
<html>
<body>

example locale | example format
-- | --
'en' | -123,456.7
'de' | -123.456,7
'fr' | -123 456,7
'hi' | -1,23,456.7
'de-CH' | -123'456.7

</body>
</html>

5. One test in the `test_table_api.py` file was updated

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: mathesar/tests/api/test_table_api.py [hunk 1]
```diff
@@ -496,9 +496,8 @@ def test_table_previews_invalid_type_cast(client, schema, engine_email_type):
         ]
     }
     response = client.post(f'/api/db/v0/tables/{table.id}/previews/', data=post_body)
-    assert response.status_code == 400
-    assert "Invalid type" in response.json()[0]['message']
-    assert "columns" in response.json()[0]['field']
+    print(response.json())
+    assert response.status_code == 500
```
Comment (dmos62) [inline mathesar/tests/api/test_table_api.py:500]: We consider 5XX responses as bugs in our code. Use an appropriate 4XX http code instead. Examine how similar errors are handled for guidance.

Also, there seems to be a left-over debugging call (line 499).

**Comment by dmos62** [comment]
@mathemancer @Jyuart conversation about regexes reminded me of this blog post about composing regex expressions using f-strings and `re.VERBOSE`: https://death.andgravity.com/f-re

**Comment by mathemancer** [comment]
> @mathemancer @Jyuart conversation about regexes reminded me of this blog post about composing regex expressions using f-strings and `re.VERBOSE`: https://death.andgravity.com/f-re

To some extent, that's the strategy taken by the function for money. The goal is to decompose into understandable nuggets, then compose it in a reasonable way:

    # An attempt to separate pieces into logical bits for easier
    # understanding and modification
    non_numeric = r"(?:[^.,0-9]+)"
    no_separator_big = r"[0-9]{4,}(?:([,.])[0-9]+)?"
    no_separator_small = r"[0-9]{1,3}(?:([,.])[0-9]{1,2}|[0-9]{4,})?"
    comma_separator_req_decimal = r"[0-9]{1,3}(,)[0-9]{3}(\.)[0-9]+"
    period_separator_req_decimal = r"[0-9]{1,3}(\.)[0-9]{3}(,)[0-9]+"
    comma_separator_opt_decimal = r"[0-9]{1,3}(?:(,)[0-9]{3}){2,}(?:(\.)[0-9]+)?"
    period_separator_opt_decimal = r"[0-9]{1,3}(?:(\.)[0-9]{3}){2,}(?:(,)[0-9]+)?"
    space_separator_opt_decimal = r"[0-9]{1,3}(?:( )[0-9]{3})+(?:([,.])[0-9]+)?"
    comma_separator_lakh_system = r"[0-9]{1,2}(?:(,)[0-9]{2})+,[0-9]{3}(?:(\.)[0-9]+)?"


    inner_number_tree = "|".join(
        [
            no_separator_big,
            no_separator_small,
            comma_separator_req_decimal,
            period_separator_req_decimal,
            comma_separator_opt_decimal,
            period_separator_opt_decimal,
            space_separator_opt_decimal,
            comma_separator_lakh_system,
        ]
    )
    inner_number_group = f"({inner_number_tree})"
    required_currency_beginning = f"{non_numeric}{inner_number_group}{non_numeric}?"
    required_currency_ending = f"{non_numeric}?{inner_number_group}{non_numeric}"
    money_finding_regex = f"^(?:{required_currency_beginning}|{required_currency_ending})$"

File: db/types/operations/cast.py [hunk 2]
```diff
@@ -962,6 +973,123 @@ def _get_text_uri_type_body_map():
...
+
+
+def _build_numeric_array_function():
+    """
+    The main reason for this function to be separate is for testing. This
+    does have some performance impact; we should consider inlining later.
+    """
+    qualified_function_name = base.get_qualified_name(NUMERIC_ARR_FUNC_NAME)
+
+    no_separator = r"[0-9]+(?:([.,])[0-9]+)?"
```
[the remark this answers is not in the document]
Comment (Jyuart) [reply db/types/operations/cast.py:1039]: Yeah, you're right. I guess I was too focused on covering all the possible cases that I also forgot to consider false positives (the cases that represent what they are not, in fact).
And still, what about examples 123,456 and 123.456 that you provided at the beginning? I guess it won't be ok just to skip them because they definitely represent numbers. Although, I understand yout concern about them not being handled correctly.

[the remark this answers is not in the document]
Comment (Jyuart) [reply db/types/operations/cast.py:1039]: @mathemancer guess you just missed this question. May I get your opinion on what we will do with numbers like `123,456` or `123.456`? And I think that such format `-123'456.7` (`de-CH`) is also not handled in the `build_mathesar_money_array` function. But should it?

[the remark this answers is not in the document]
Comment (mathemancer) [reply db/types/operations/cast.py:1039]: Ah, sorry. I really wish GH would highlight new comments somehow.

For the moment, we should not cast `123,456` or `123.456`. While those _are_ numbers, casting in the context of inference has the potential to irreversibly corrupt data as described above. We'll eventually be getting more and more sophisticated about these things, and should be able to add those later. If the user wants to cast a column with such values, they'll have to do some manual work removing commas or adding `.0` or `,0` to each value.

As for `de-CH`, you're right; good catch! Please add that to this PR, and raise a bug for the money type.

File: mathesar/tests/api/test_table_api.py (hunk 1 above)
Comment (Jyuart) [reply mathesar/tests/api/test_table_api.py]: The test itself changed, so it was fixed another way

File: db/types/categories.py [hunk 3]
```diff
@@ -72,6 +71,7 @@
 NUMERIC_TYPES = frozenset({
     *INTEGER_TYPES,
     *DECIMAL_TYPES,
+    PostgresType.NUMERIC
```
Comment (mathemancer) [inline db/types/categories.py:74]: Why this change? `NUMERIC` is a decimal type.

**Review by mathemancer — APPROVED** [review]
Okay, I think this is almost there. I did have a question about a change you introduced, but I don't think it should break anything.

File: db/types/categories.py (hunk 3 above)
Comment (Jyuart) [reply db/types/categories.py:74]: My intention was to exclude it from being used in the `_get_decimal_number_type_body_map` function which creates 'default' casting function for other decimal types but to still treat it as a numeric type.
